### ISSUE TITLE
semicolon_if_nothing_returned lint when future type inference fails with the semicolon

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust-clippy#9729]
### Summary

Using `semicolon_if_nothing_returned` can lead to clippy complaining that there ought to be a semicolon, in a location where placing one causes type inference to fail.

### Lint Name

semicolon_if_nothing_returned

### Reproducer

I tried this code:

```rust
#![warn(clippy::semicolon_if_nothing_returned)]

pub async fn something() {
    if true {
        eprintln!("hi");
    } else {
        futures::future::pending().await
    }
}
```

I saw this happen:

```
7 |         futures::future::pending().await
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `futures::future::pending().await;`
```

Adopting the suggestion causes:

```
7 |         futures::future::pending().await;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T` declared on the function `pending`
```

I expected to see this happen:

No complaint.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=805cabba28a97e36a701643678ad26fb

### Version

```text
Rust 1.64.0 stable on the playground.
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
